addons/binding/org.openhab.binding.evohome/src/main/java/org/openhab/binding/evohome/internal/api/ApiAccess.java

```diff
@@ -0,0 +1,198 @@
+
+    private Authentication authenticationData;
+    private String applicationId;
+
+    public ApiAccess(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    /**
+     * Sets the authentication details on the type
+     *
+     * @param authentication The authentication details to apply
+     */
+    public void setAuthentication(Authentication authentication) {
+        authenticationData = authentication;
+    }
+
+    /**
+     * Gets the current authentication details of the type
+     *
+     * @return The current authentication details
+     */
+    public Authentication getAuthentication() {
+        return authenticationData;
+    }
+
+    /**
+     * Sets the application id on the type
+     *
+     * @param applicationId The application id to apply
+     */
+    public void setApplicationId(String applicationId) {
+        this.applicationId = applicationId;
+    }
+
+    /**
+     * Issues an HTTP request on the API's URL. Makes sure that the request is correctly formatted.
+     *
+     * @param method The HTTP method to use (POST, GET, ...)
+     * @param url The URL to query
+     * @param headers The optional additional headers to apply, can be null
+     * @param requestData The optional request data to use, can be null
+     * @param contentType The content type to use with the request data. Required when using requestData
+     * @param out An instance of the return type, ignored when value is null
+     * @return The result of the request or null
+     */
+    public <TOut> TOut doRequest(HttpMethod method, String url, Map<String, String> headers, String requestData,
+            String contentType, Class<TOut> outClass, TOut out) {
+
+        TOut retVal = out;
+        logger.debug("Requesting: [{}]", url);
+
+        try {
+            Request request = httpClient.newRequest(url).method(method);
+
+            if (headers != null) {
+                for (Map.Entry<String, String> header : headers.entrySet()) {
+                    request.header(header.getKey(), header.getValue());
+                }
+            }
+
+            if (requestData != null) {
+                request.content(new StringContentProvider(requestData), contentType);
+            }
+
+            ContentResponse response = request.send();
+
+            logger.debug("Response: {}", response);
+            logger.trace("\n{}\n{}", response.getHeaders(), response.getContentAsString());
+
+            if ((response.getStatus() == HttpStatus.OK_200) || (response.getStatus() == HttpStatus.ACCEPTED_202)) {
+                String reply = response.getContentAsString();
+
+                if (outClass != null && out != null) {
+                    retVal = new Gson().fromJson(reply, outClass);
+                }
+            } else {
+                retVal = null;
+            }
+        } catch (TimeoutException | InterruptedException | ExecutionException e) {
+            logger.error("Error in handling request", e);
```

Comment:
A timeout is clearly an expected path and it should merely set the Thing status to OFFLINE.
Logging errors must only be done if this is a bug situation that should be reported by users.

Reply:
You mean timeouts happen, even if everything works as expected? Sounds weird...
My point was that if you e.g. do not have a network connection, I would expect a TimeoutException here - and in that case, the Thing must be set to OFFLINE. Or is this already happening at some other place?
For the other two exceptions in that catch clause, you can choose error or warning as you like - assuming it is something that must normally never happen.

Reply:
> You mean timeouts happen, even if everything works as expected?
Yeah, or my internet connection is unstable, but I see at least one timeout per day on requesting status. I could make it better by queueing up requests and implement a retry mechanism and throw a TimeoutException if happens `RetryCount` times, but that seems like something for a future version. For now, I don't really care if a status update fails. Then again, maybe I should solve it there then (only on update); let's go for that then.

Reply:
Setting the Thing to OFFLINE (and back to ONLINE after the next successful update) would at least give a hint to the user that something might not be fully stable&reliable.

Reply:
Makes sense, I'll take that along